src/rpc/notifications.cpp

```diff
@@ -0,0 +1,381 @@
+
+        LOCK(m_cs_notification_queue);
+        m_registered_notification_types.insert(type);
+    }
+
+    void unregisterType(NotificationType type)
+    {
+        LOCK(m_cs_notification_queue);
+        m_registered_notification_types.erase(type);
+    }
+
+    void unregisterAllTypes()
+    {
+        LOCK(m_cs_notification_queue);
+        m_registered_notification_types.clear();
+    }
+
+    void addToQueue(NotificationEntry entry)
+    {
+        LOCK(m_cs_notification_queue);
+
+        size_t queueSize = m_queue.size();
+        if (queueSize > MAX_QUEUE_SIZE) {
+            m_queue.pop_front();
+            LogPrintf("RPC Notification limit has been reached, dropping oldest element\n");
+        }
+        m_map_sequence_numbers[entry.m_type]++;
+        entry.m_sequence_number = m_map_sequence_numbers[entry.m_type];
+        m_queue.push_back(entry);
+    }
+
+    /* checks if a certain notification type is registered */
+    bool isTypeRegistered(NotificationType type)
+    {
+        LOCK(m_cs_notification_queue);
+        return (m_registered_notification_types.find(type) != m_registered_notification_types.end());
+    }
+};
+
+class NotificationQueueManager : public CValidationInterface
+{
+public:
+    CCriticalSection m_cs_queue_manager;
+    std::map<clientUUID_t, NotificationQueue*> m_map_sequence_numbers;
+
+    NotificationQueue* getQueue(const clientUUID_t& clientid)
+    {
+        LOCK(m_cs_queue_manager);
+        return m_map_sequence_numbers[clientid];
+    }
+
+    NotificationQueue* addQueue(const clientUUID_t& clientid)
+    {
+        LOCK(m_cs_queue_manager);
+        m_map_sequence_numbers[clientid] = new NotificationQueue();
+        return m_map_sequence_numbers[clientid];
+    }
+
+    void NotifyTransaction(const CTransactionRef& ptx)
+    {
+        LOCK(m_cs_queue_manager);
+        for (auto& queueEntry : m_map_sequence_numbers) {
+            if (!queueEntry.second->isTypeRegistered(NotificationType::Tx)) continue;
+
+            NotificationEntry entry;
+            entry.m_type = NotificationType::Tx;
+            entry.m_notification.setStr(ptx->GetHash().GetHex());
+            queueEntry.second->addToQueue(entry);
+        }
+    }
+
+    void BlockConnected(const std::shared_ptr<const CBlock>& pblock, const CBlockIndex* pindexConnected, const std::vector<CTransactionRef>& vtxConflicted) override
+    {
+        for (const CTransactionRef& ptx : pblock->vtx) {
+            // Do a normal notify for each transaction added in the block
+            NotifyTransaction(ptx);
+        }
+    }
+
+    void BlockDisconnected(const std::shared_ptr<const CBlock>& pblock) override
+    {
+        for (const CTransactionRef& ptx : pblock->vtx) {
+            // Do a normal notify for each transaction removed in block disconnection
+            NotifyTransaction(ptx);
+        }
+    }
+
+    void UpdatedBlockTip(const CBlockIndex* pindexNew, const CBlockIndex* pindexFork, bool fInitialDownload) override
```

Comment:
I dont think this is the notification we want  here - dont we want to use BlockConnected/Disconnected to notify clients of all connected blocks, not just new tips after reorgs which potentially connect multiple blocks?